spring-integration-file/src/main/java/org/springframework/integration/file/filters/AbstractMarkerFilePresentFileListFilter.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.integration.file.filters;
+
+import java.lang.reflect.Array;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/**
+ * A FileListFilter that only passes files matched by one or more {@link FileNameFileListFilter}
+ * if a corresponding marker file is also present to indicate a file transfer is complete.
+ *
+ * @author Gary Russell
+ * @since 5.0
+ *
+ */
+public abstract class AbstractMarkerFilePresentFileListFilter<F> implements FileListFilter<F> {
+
+	private final Map<FileNameFileListFilter<F>, Function<String, String>> filtersAndFunctions = new HashMap<>();
+
+	/**
+	 * Construct an instance with a single {@link FileNameFileListFilter} and a suffix that will
+	 * be used to replace the file suffix of matched files when looking for the marker file.
+	 * i.e. if a file {@code foo.txt} is matched by the filter and the suffix is "complete",
+	 * this filter will only pass "foo.txt" if "foo.complete" is present.
```

Comment:
M-m-m. Why do you select the extension replacing approach as default one?
Seems for me `foo.txt.complete` is much robust solution, especially when we may have `foo.exe`, `foo.jpg`, `foo.java` at the same folder.